File: apps/frontend/src/services/quality.ts
import api, { ApiResponse, unwrap } from "./api";

export interface DataQualityIssue {
  issue_type: string;
  severity: string;
  message: string;
  entry_id?: string;
  fields?: Record<string, unknown>;
  detected_at: string;
}

export interface DataQualitySummary {
  total: number;
  by_severity: Record<string, number>;
  by_type: Record<string, number>;
}

export interface DataQualityReport {
  issues: DataQualityIssue[];
  summary: DataQualitySummary;
  generated_at: string;
}

export const fetchDataQualityReport = async (params?: {
  state?: string;
  district?: string;
  mandi?: string;
  commodity?: string;
  date_from?: string;
  date_to?: string;
}) => {
  const response = await api.get<ApiResponse<DataQualityReport>>("/quality/mandi", { params });
  return unwrap(response.data);
};

export const fetchQualityIssues = async () => {
  const response = await api.get<ApiResponse<unknown[]>>("/quality/issues");
  return unwrap(response.data);
};

File: apps/backend/app/schemas/quality.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel


class DataQualityIssue(BaseModel):
    issue_type: str
    severity: str
    message: str
    entry_id: Optional[str] = None
    fields: Optional[dict] = None
    detected_at: datetime


class DataQualitySummary(BaseModel):
    total: int
    by_severity: dict
    by_type: dict


class DataQualityReport(BaseModel):
    issues: List[DataQualityIssue]
    summary: DataQualitySummary
    generated_at: datetime
